signer/src/network/in_memory.rs

```diff
@@ -0,0 +1,73 @@
+//! # In-memory signer network client.
+//!
+//! The client itself is a thin wrapper over a tokio broadcast
+//! channel, with deduplication logic to prevent a single client
+//! from receiving it's own messages.
+
+use std::collections::VecDeque;
+
+use tokio::sync::broadcast;
+
+pub const BROADCAST_CHANNEL_CAPACITY: usize = 10_000;
+
+#[derive(Debug)]
+pub struct Client {
+    sender: broadcast::Sender<super::Msg>,
+    receiver: broadcast::Receiver<super::Msg>,
+    recently_sent: VecDeque<super::Msg>,
+}
+
+/// In-memory communication network
+#[derive(Debug)]
+pub struct Network {
+    sender: broadcast::Sender<super::Msg>,
+}
+
+impl Network {
+    /// Construct a new in-memory communication entwork
+    pub fn new() -> Self {
+        let (sender, _) = broadcast::channel(BROADCAST_CHANNEL_CAPACITY);
+        Self { sender }
+    }
+
+    /// Connect a new signer to this network
+    pub fn connect(&self) -> Client {
+        let sender = self.sender.clone();
+        let receiver = sender.subscribe();
+        let recently_sent = VecDeque::new();
+
+        Client {
+            sender,
+            receiver,
+            recently_sent,
+        }
+    }
+}
+
+impl super::Client for Client {
+    type Error = Error;
+    async fn broadcast(&mut self, msg: super::Msg) -> Result<(), Self::Error> {
+        self.recently_sent.push_back(msg.clone());
+        self.sender.send(msg)?;
+        Ok(())
+    }
+
+    async fn receive(&mut self) -> Result<super::Msg, Self::Error> {
+        let mut msg = self.receiver.recv().await?;
+
+        while Some(&msg) == self.recently_sent.get(0) {
+            self.recently_sent.pop_front();
+            msg = self.receiver.recv().await?;
+        }
+
+        Ok(msg)
+    }
```

Comment:
Nice, I see. This is a simple network that works. Everyone receives everyone else's messages and I also receive my own messages. 👍🏽 

An alternative approach, is to give each message a unique ID and filter on IDs that you have broadcasted. That way you do not need to clone the message and you get (roughly) the same guarantees. The approach here is fine too.

Reply:
Yeah, this is a bit of an ugly hack to prevent the signers own messages from echoing back. I like the idea of using IDs instead of matching on the entire message. I guess being able to identify messages will come in handy in the future too. I'll update.

Reply:
Updated.